File: hdf-diff/go/render/terminal.go
package render

import (
	"fmt"
	"strings"

	diff "github.com/mitre/hdf-diff/go"
)

// ANSI color codes.
const (
	ansiReset  = "\x1b[0m"
	ansiGreen  = "\x1b[32m"
	ansiRed    = "\x1b[31m"
	ansiYellow = "\x1b[33m"
	ansiBold   = "\x1b[1m"
	ansiDim    = "\x1b[2m"
)

// Terminal renders an HdfComparison for terminal display with optional ANSI colors.
//
//   - DetailSummary: header + summary line only
//   - DetailControl: header + requirement list (excludes unchanged) + summary
//   - DetailFull: header + all requirements including unchanged with changeReasons/fieldChanges + summary
func Terminal(comparison diff.HdfComparison, opts Options) (string, error) {
	detail := opts.effectiveDetail()
	useColor := !opts.NoColor

	var sb strings.Builder

	// Header
	sb.WriteString(buildHeaderLine(comparison, useColor))
	sb.WriteString("\n")

	if detail == DetailSummary {
		sb.WriteString("\n")
		sb.WriteString(buildSummaryLine(comparison, useColor))
		return sb.String(), nil
	}

	sb.WriteString("\n")

	// Requirement lines
	filtered := filterRequirements(comparison.RequirementDiffs, opts)

	for _, req := range filtered {
		// In control mode, skip unchanged requirements
		if detail == DetailControl && req.State == diff.StateUnchanged {
			continue
		}
		sb.WriteString(renderRequirementLine(req, detail, useColor))
		sb.WriteString("\n")
	}

	sb.WriteString("\n")
	sb.WriteString(buildSummaryLine(comparison, useColor))

	return sb.String(), nil
}

// symbolAndColor returns the terraform-style symbol and color wrapper for a state.
func symbolAndColor(state diff.RequirementState, useColor bool) (string, func(string) string) {
	identity := func(s string) string { return s }

	switch state {
	case diff.StateFixed, diff.StateNew:
		if useColor {
			return "+", func(s string) string { return ansiGreen + s + ansiReset }
		}
		return "+", identity
	case diff.StateRegressed, diff.StateAbsent:
		if useColor {
			return "-", func(s string) string { return ansiRed + s + ansiReset }
		}
		return "-", identity
	case diff.StateUpdated:
		if useColor {
			return "~", func(s string) string { return ansiYellow + s + ansiReset }
		}
		return "~", identity
	case diff.StateUnchanged, diff.StateMoved, diff.StateSplit, diff.StateMerged:
		if useColor {
			return " ", func(s string) string { return ansiDim + s + ansiReset }
		}
		return " ", identity
	default:
		if useColor {
			return " ", func(s string) string { return ansiDim + s + ansiReset }
		}
		return " ", identity
	}
}

// renderRequirementLine formats a single requirement for terminal display.
func renderRequirementLine(req diff.RequirementDiff, detail DetailLevel, useColor bool) string {
	symbol, colorFn := symbolAndColor(req.State, useColor)
	transition := formatStatusTransition(req)

	line := fmt.Sprintf("  %s %s  %s    %s", symbol, req.ID, req.Title, transition)

	if detail == DetailFull {
		if len(req.ChangeReasons) > 0 {
			reasons := make([]string, len(req.ChangeReasons))
			for i, r := range req.ChangeReasons {
				reasons[i] = string(r)
			}
			line += fmt.Sprintf("  [%s]", strings.Join(reasons, ", "))
		}
		fc := formatFieldChangesWithArrow(req.FieldChanges, "\u2192")
		if fc != "" {
			line += "  " + fc
		}
	}

	return colorFn(line)
}

// formatStatusTransition formats the status change portion of a requirement line.
func formatStatusTransition(req diff.RequirementDiff) string {
	if req.State == diff.StateNew {
		return "(new)"
	}
	if req.State == diff.StateAbsent {
		return "(absent)"
	}

	var parts []string

	if req.OldEffectiveStatus != "" && req.NewEffectiveStatus != "" {
		parts = append(parts, fmt.Sprintf("%s \u2192 %s", req.OldEffectiveStatus, req.NewEffectiveStatus))
	}

	if req.State != diff.StateUnchanged {
		parts = append(parts, fmt.Sprintf("(%s)", req.State))
	}

	return strings.Join(parts, "   ")
}

// buildHeaderLine builds the header for terminal output.
func buildHeaderLine(comparison diff.HdfComparison, useColor bool) string {
	header := fmt.Sprintf("HDF Comparison: %s", comparison.ComparisonMode)

	// Find old and new source timestamps
	var oldTimestamp, newTimestamp string
	for _, s := range comparison.Sources {
		switch s.Role {
		case diff.RoleOld, diff.RoleGolden, diff.RoleReference:
			if oldTimestamp == "" {
				oldTimestamp = s.AssessmentTimestamp
			}
		case diff.RoleNew, diff.RoleSystem:
			if newTimestamp == "" {
				newTimestamp = s.AssessmentTimestamp
			}
		}
	}

	if oldTimestamp != "" && newTimestamp != "" {
		oldDate := strings.SplitN(oldTimestamp, "T", 2)[0]
		newDate := strings.SplitN(newTimestamp, "T", 2)[0]
		header += fmt.Sprintf(" (%s \u2192 %s)", oldDate, newDate)
	}

	if useColor {
		return ansiBold + header + ansiReset
	}
	return header
}

// buildSummaryLine builds the summary line for terminal output.
func buildSummaryLine(comparison diff.HdfComparison, useColor bool) string {
	s := comparison.Summary
	line := fmt.Sprintf("Summary: %d fixed, %d regressed, %d new, %d absent, %d unchanged, %d updated (%d total)",
		s.Fixed, s.Regressed, s.New, s.Absent, s.Unchanged, s.Updated, s.Total)

	if useColor {
		return ansiBold + line + ansiReset
	}
	return line
}

File: hdf-diff/src/renderers/terminal.ts
import type { HDFComparison, RequirementDiff } from '../types.js';
import type { RenderOptions } from './types.js';
import { filterRequirements } from './filter.js';

// ANSI color codes
const RESET = '\x1b[0m';
const GREEN = '\x1b[32m';
const RED = '\x1b[31m';
const YELLOW = '\x1b[33m';
const BOLD = '\x1b[1m';
const DIM = '\x1b[2m';

/**
 * Get the symbol and color for a requirement state.
 */
function getSymbolAndColor(
  state: string,
  useColor: boolean,
): { symbol: string; colorFn: (s: string) => string } {
  const identity = (s: string): string => s;

  switch (state) {
    case 'fixed':
    case 'new':
      return {
        symbol: '+',
        colorFn: useColor ? (s) => `${GREEN}${s}${RESET}` : identity,
      };
    case 'regressed':
    case 'absent':
      return {
        symbol: '-',
        colorFn: useColor ? (s) => `${RED}${s}${RESET}` : identity,
      };
    case 'updated':
      return {
        symbol: '~',
        colorFn: useColor ? (s) => `${YELLOW}${s}${RESET}` : identity,
      };
    case 'unchanged':
    default:
      return {
        symbol: ' ',
        colorFn: useColor ? (s) => `${DIM}${s}${RESET}` : identity,
      };
  }
}

/**
 * Format the status transition for a requirement.
 */
function formatStatusTransition(req: RequirementDiff): string {
  if (req.state === 'new') {
    return '(new)';
  }
  if (req.state === 'absent') {
    return '(absent)';
  }

  const parts: string[] = [];

  // Show status change if both exist
  if (req.oldEffectiveStatus && req.newEffectiveStatus) {
    parts.push(`${req.oldEffectiveStatus} → ${req.newEffectiveStatus}`);
  }

  // Add state label for non-obvious transitions
  if (req.state !== 'unchanged') {
    parts.push(`(${req.state})`);
  }

  return parts.join('   ');
}

/**
 * Format field changes for full detail.
 */
function formatFieldChangesForTerminal(req: RequirementDiff): string {
  if (req.fieldChanges.length === 0) return '';

  const parts = req.fieldChanges.map((fc) => {
    if (fc.op === 'add') return `+${fc.path}: ${JSON.stringify(fc.newValue)}`;
    if (fc.op === 'remove') return `-${fc.path}: ${JSON.stringify(fc.oldValue)}`;
    return `${fc.path}: ${JSON.stringify(fc.oldValue)} → ${JSON.stringify(fc.newValue)}`;
  });

  return parts.join('; ');
}

/**
 * Render a single requirement line.
 */
function renderRequirementLine(
  req: RequirementDiff,
  detail: string,
  useColor: boolean,
): string {
  const { symbol, colorFn } = getSymbolAndColor(req.state, useColor);
  const title = req.title ?? '';
  const transition = formatStatusTransition(req);

  let line = `  ${symbol} ${req.id}  ${title}    ${transition}`;

  if (detail === 'full') {
    // Add change reasons
    if (req.changeReasons.length > 0) {
      line += `  [${req.changeReasons.join(', ')}]`;
    }
    // Add field changes
    const fieldChanges = formatFieldChangesForTerminal(req);
    if (fieldChanges) {
      line += `  ${fieldChanges}`;
    }
  }

  return colorFn(line);
}

/**
 * Build the summary line.
 */
function buildSummaryLine(comparison: HDFComparison, useColor: boolean): string {
  const { summary } = comparison;

  const parts = [
    `${summary.fixed} fixed`,
    `${summary.regressed} regressed`,
    `${summary.new} new`,
    `${summary.absent} absent`,
    `${summary.unchanged} unchanged`,
    `${summary.updated} updated`,
    `(${summary.total} total)`,
  ];

  const line = `Summary: ${parts.join(', ')}`;

  if (useColor) {
    return `${BOLD}${line}${RESET}`;
  }
  return line;
}

/**
 * Build the header line.
 */
function buildHeaderLine(comparison: HDFComparison, useColor: boolean): string {
  const mode = comparison.comparisonMode;
  const oldSource = comparison.sources.find((s) => s.role === 'old' || s.role === 'golden' || s.role === 'reference');
  const newSource = comparison.sources.find((s) => s.role === 'new' || s.role === 'system');

  let header = `HDF Comparison: ${mode}`;

  const oldTimestamp = oldSource?.assessmentTimestamp;
  const newTimestamp = newSource?.assessmentTimestamp;

  if (oldTimestamp && newTimestamp) {
    // Format as date-only if possible, otherwise use full timestamp
    // split() always returns at least one element, so [0] is safe
    const oldDate = oldTimestamp.split('T')[0]!;
    const newDate = newTimestamp.split('T')[0]!;
    header += ` (${oldDate} → ${newDate})`;
  }

  if (useColor) {
    return `${BOLD}${header}${RESET}`;
  }
  return header;
}

/**
 * Render an HDFComparison for terminal display with optional ANSI colors.
 *
 * - `detail: 'summary'` -- just the summary line
 * - `detail: 'control'` -- requirement list + summary (excludes unchanged)
 * - `detail: 'full'`    -- all requirements including unchanged, with changeReasons and fieldChanges
 *
 * When `color: false`, no ANSI escape codes are emitted.
 * Default detail level: `'control'`. Default color: `true`.
 */
export function renderTerminal(
  comparison: HDFComparison,
  options?: RenderOptions,
): string {
  const detail = options?.detail ?? 'control';
  const useColor = options?.color ?? true;

  const lines: string[] = [];

  // Header
  lines.push(buildHeaderLine(comparison, useColor));
  lines.push('');

  if (detail === 'summary') {
    lines.push(buildSummaryLine(comparison, useColor));
    return lines.join('\n');
  }

  // Requirement lines
  const filtered = filterRequirements(comparison.requirementDiffs, options);

  for (const req of filtered) {
    // In 'control' mode, skip unchanged requirements
    if (detail === 'control' && req.state === 'unchanged') {
      continue;
    }
    lines.push(renderRequirementLine(req, detail, useColor));
  }

  lines.push('');
  lines.push(buildSummaryLine(comparison, useColor));

  return lines.join('\n');
}
